Add length validation for rename_replacement parameter in snapshot restore request

gmjehovich opened the pull request:
## Summary
Adds validation to the `rename_replacement` parameter in the snapshot restore API to prevent resource exhaustion.

## Changes
- Added validation in `RestoreSnapshotRequest.validate()` to limit `rename_replacement` UTF-8 byte length to `MAX_INDEX_NAME_BYTES` (255 bytes)
- Added integration test `testRenameReplacementTooLongRejected()` to verify the validation works correctly
- Imported `MAX_INDEX_NAME_BYTES` from `MetadataCreateIndexService` to maintain consistency with index name constraints

## Testing
- Integration test confirms validation rejects overly long values
- Test confirms valid values (≤255 bytes) are still accepted
- Verified existing restore functionality is not affected

gmjehovich committed:
Add length validation for rename_replacement parameter in snapshot restore

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Merge branch 'main' into fix-restore-snapshot-oom

gmjehovich committed:
Update RestoreSnapshotIT.java

gmjehovich committed:
Merge branch 'main' into fix-restore-snapshot-oom

gmjehovich committed:
Merge branch 'main' into fix-restore-snapshot-oom

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

gmjehovich committed:
Update docs/changelog/137859.yaml

ywangd commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java:
```diff
@@ -152,6 +154,20 @@ public ActionRequestValidationException validate() {
+                        + "] bytes",
+                    validationException
+                );
+            }
```
Due to usage of back-references and escapes, the result index name can be shorter than 255 when `renameReplacement` is longer than 255. So we could be preventing some legal index names in edge cases. I think it might be OK though since such long index names are very rare.

ywangd replied:
Yeah. We could do something along that line. I'd prefer to be a bit more strict, e.g. allowing an extra buffer of 10 chars.

ywangd reviewed (commented):
Looking good. I have a suggestion about whether we should further restrict building unnecessarily long strings for renamed index names.

gmjehovich committed:
Fixup integ testing

# Conflicts:
#	server/src/internalClusterTest/java/org/elasticsearch/snapshots/RestoreSnapshotIT.java

gmjehovich committed:
Fix integration tests

gmjehovich committed:
Add new IT test case for long replacement

gmjehovich commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/RestoreSnapshotIT.java:
```diff
@@ -1102,4 +1102,63 @@ public void testExplainUnassigableDuringRestore() {
+            .setRenamePattern("b+")
+            .setRenameReplacement("a".repeat(255))
+            .setWaitForCompletion(true)
+            .get();
```
I've added a test that verifies the 255×255 scenario fails with proper index name validation rather than OOM.

Regarding the manual loop implementation similar to `safeReplace`, I attempted this approach but ran into issues with accurately estimating sizes for back-references ($1, $2, etc.). The conservative estimation (assuming each group could expand to the full match size) would reject valid cases where capture groups only match portions of the text. These false rejections would likely be edge cases (complex regex patterns with multiple capture groups), but they're still valid operations that would be blocked.

For now, I've kept the simpler validation that prevents the unbounded OOM cases. While the 65KB intermediate strings aren't ideal, they're bounded, and I don't believe they'll cause an OOM error (although I haven't actually tested it with 50 concurrent requests). As you noted, the heap usage is nowhere near the unbounded case this PR fixes.

How much of a concern is the current `replaceAll` implementation for you? The manual loop would be more efficient, but with the tradeoff that some valid rename patterns may be rejected. Happy to implement it if you think the efficiency gain is worth that tradeoff, or it could be explored more in a follow-up PR.

ywangd replied:
> I attempted this approach but ran into issues with accurately estimating sizes for back-references ($1, $2, etc.).
> with the tradeoff that some valid rename patterns may be rejected.

I don't quite follow. I was thinking something like the following change to RestoreService.java

```diff
@@ -1074,7 +1075,22 @@ public final class RestoreService implements ClusterStateApplier {
         if (prefix != null) {
             index = index.substring(prefix.length());
         }
-        renamedIndex = index.replaceAll(request.renamePattern(), request.renameReplacement());
+        final var matcher = Pattern.compile(request.renamePattern()).matcher(index);
+        var found = matcher.find();
+        if (found) {
+            final var sb = new StringBuilder();
+            do {
+                matcher.appendReplacement(sb, request.renameReplacement());
+                found = matcher.find();
+            } while (found && sb.length() <= 255);
+            if (sb.length() > 255) {
+                throw new IllegalArgumentException();
+            }
+            matcher.appendTail(sb);
+            renamedIndex = sb.toString();
+        } else {
+            renamedIndex = index;
+        }
         if (prefix != null) {
             renamedIndex = prefix + renamedIndex;
         }
```

Does it not work and could reject valid cases? 

> it could be explored more in a follow-up PR

I am OK for it to be a follow-up if you prefer.

gmjehovich replied:
Ahhh I see what you mean now, I think this avoids the issue I was running into... I was trying to copy the logic in [safeReplace](https://github.com/elastic/elasticsearch/blob/f67382272370b8d722b06d9480fc676be2a07588/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Replace.java#L174-L186) too closely, which does an estimation before actually making the string and led to that issue I described. I don't think we need to be that careful here, and this approach you provided should work without rejecting valid cases. 

I will do some testing on it tomorrow.

gmjehovich committed:
Add buffer to rename_replacement size check; add safeRenameIndex()

gmjehovich committed:
Randomize rename_replacement in testRenameReplacementNameTooLong()

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Merge branch 'main' into fix-restore-snapshot-oom

ywangd reviewed (commented):
This looks great. I only had one minor comment. Thanks!

gmjehovich committed:
Add unit test for safeRenameIndex()

gmjehovich committed:
fix testRenameReplacementNameTooLong() to use createTestInstance()

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Merge branch 'main' into fix-restore-snapshot-oom

ywangd reviewed (approved):
LGTM

Thanks for the iterations!
